### PR TITLE
[api-extractor] Generate documentation for setters without getters

### PR DESCRIPTION
**Pull request by zelliott**
## Summary

Instead of ignoring setters entirely, this PR changes API Extractor to generate documentation for a setter if and only if there is no associated getter. If there is an associated getter, the setter is ignored. Note that API Extractor _still_ has the `ae-missing-getter` warning for teams that want to discourage setters without getters.

In the future, perhaps an `isWriteonly` bit can be added to these nodes, similar to https://github.com/microsoft/rushstack/issues/3265.

## Details

N/A

## How it was tested

Added a relevant scenario to the api-documenter-test build test, ran `rush rebuild`, manually verified that the generated artifacts were correct.

### CODE REVIEW

File: build-tests/api-documenter-test/config/api-extractor.json [hunk 1]
```diff
@@ -18,5 +18,14 @@
...
 
-  "testMode": true
+  "testMode": true,
+
+  "messages": {
+    "extractorMessageReporting": {
+      // Purposefully disabled for the `writeonlyProperty` test case.
+      "ae-missing-getter": {
+        "logLevel": "none"
+      }
```
Comment (zelliott) [inline build-tests/api-documenter-test/config/api-extractor.json:28]: I'd prefer to "expect" that there's a single `ae-missing-getter` warning - is that easy to do with how these build tests are set up?

Comment (octogonz) [reply build-tests/api-documenter-test/config/api-extractor.json:28]: No, there isn't an actual test harness here - this is a normal project without any assertions; we just compare its output against Git what's in.

Actually most of the test cases have been moved into the [build-tests/api-extractor-scenarios](https://github.com/microsoft/rushstack/tree/main/build-tests/api-extractor-scenarios) project, which invokes AE separately for each test, and allows each test to have its own `api-extractor-overrides.json` that overrides the default `api-extractor.json` config.  That project doesn't support API doc generation though.

But given that the point of `api-documenter-test` is to invoke API Documenter (versus testing API Extractor edge cases), I'm fine with suppressing `ae-missing-getter` in this project. 👍

File: apps/api-extractor/src/generators/ApiModelGenerator.ts [hunk 2]
```diff
@@ -817,13 +821,17 @@ export class ApiModelGenerator {
...
-        astDeclaration.declaration as ts.PropertyDeclaration;
-
       const nodesToCapture: IExcerptBuilderNodeToCapture[] = [];
 
       const propertyTypeTokenRange: IExcerptTokenRange = ExcerptBuilder.createEmptyTokenRange();
-      nodesToCapture.push({ node: propertyDeclaration.type, tokenRange: propertyTypeTokenRange });
+
+      // If the property declaration's type is `undefined`, then we're processing a setter with no corresponding
+      // getter. Use the parameter type instead (note that TSC will enforce that the setter has exactly one
+      // parameter).
```
Comment (octogonz) [inline apps/api-extractor/src/generators/ApiModelGenerator.ts:830]: > note that TSC will enforce that the setter has exactly one parameter

The compiler might enforce it, but we should validate that assumption here.
For example, what if the .d.ts file was manually authored, or what if some new syntax
variant is introduced in a future?

Comment (octogonz) [reply apps/api-extractor/src/generators/ApiModelGenerator.ts:830]: Some [experimentation shows](https://www.typescriptlang.org/play?strict=false&noImplicitAny=false&strictNullChecks=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false#code/MYGwhgzhAEAaBQBve1XQOYFMAu0D2A1mAJ4AUAlNMmjdAE44CudAdgNwpoC+nqEO+ImQBulRD17R+uFnmwB5IRSo8eQA) that it's a compile error (TS 1049) if the setter does not have exactly one parameter.  However the compiler happily accepts setters/getters with missing type annotations:

```ts
class X
{
    get okay() {
        return;
    }
    set okay(v) {}

    set notOkay() {} // error TS 20149
}
```

Comment (octogonz) [reply apps/api-extractor/src/generators/ApiModelGenerator.ts:830]: ...but looking more closely at the code, the `IExcerptBuilderNodeToCapture.node` allows `undefined`.  So your code is functionally correct, we just should eliminate the `!`.

Comment (octogonz) [reply apps/api-extractor/src/generators/ApiModelGenerator.ts:830]: I pushed a small fix

**Review by octogonz — APPROVED** [review]

**Comment by octogonz** [comment]
Released with API Extractor 7.24.0
